virtual-programs/camera-rpi.folk

```diff
@@ -0,0 +1,303 @@
+if {$::isLaptop} return
+
+set makeCamera {
+    set cpp [C++]
+    $cpp include <iostream>
+    $cpp include <iomanip>
+    $cpp include <mutex>
+    $cpp include <queue>
+    $cpp include <sys/mman.h>
+
+    $cpp include <libcamera/libcamera.h>
+    # osnr: HACK: just throwing any possible path in.
+    $cpp cflags -I/usr/local/include/libcamera -I/usr/include/libcamera
+    c loadlibLd libcamera
+
+    defineImageType $cpp
+    $cpp code {
+        using namespace libcamera;
+
+        std::unique_ptr<CameraManager> cm;
+        std::shared_ptr<Camera> camera;
+	std::unique_ptr<CameraConfiguration> config;
+	FrameBufferAllocator *allocator;
+
+        // This vector always owns all the request objects.
+	std::vector<std::unique_ptr<Request>> requests;
+
+        std::mutex completedRequestsMutex;
+        std::queue<Request *> completedRequests;
+
+        uint32_t frameWidth;
+        uint32_t frameHeight;
+        uint32_t frameBytesPerRow;
+
+        static void requestComplete(Request *request);
+    }
+    $cpp proc cameraOpen {char* id int width int height} void {
+        cm = std::make_unique<CameraManager>();
+        cm->start();
```

Comment:
Is there a place you "should" be running `cm->stop()`? There is something really similar with libfreenect2 and if I don't run stop it will keep the camera on even after the process is killed.

I know it's not a huge issue here but it would solve my problem if we could find a way to gracefully clean up processes when ending, maybe using a statement like you mentioned in the PR description.

Reply:
Probably in an `On unmatch` in the subprocess, I think? It might get killed before that can execute, but we probably want to have some hook like that that gets a few milliseconds to do cleanup.